### PR TITLE
gsoc-2024: Add work product for mimalloc

### PR DESCRIPTION
**Pull request by huyang531**

### CODE REVIEW

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# Unikraft GSoC'24: Synchronization Support In Internal Libraries
+
+<img src="https://unikraft.org/images/unikraft-gsoc24.png"/>
+
+## Summary
+
+Memory allocators have a significant impact on application performance.
+There have been some [research papers](https://dl.acm.org/doi/10.1145/378795.378821) which compared different memory allocators.
+According to their work, Switching to the appropiate memory allocator may improve the application performance by 60%.
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
According to their work, switching to the appropiate memory allocator may improve the application performance by 60%.
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+Unikraft currently supports only one non-synchronized memory allocator the *binary buddy* allocator.
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
Unikraft currently supports only one non-synchronized memory allocator, the *binary buddy* allocator.
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+So, this project aims to enhance synchronization support in Unikraft's internal libraries and port [*mimalloc*](https://github.com/microsoft/mimalloc) (pronounced "me-malloc"), a high-performance general-purpose memory allocator developed by Microsoft, to it.
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
Therefore, this project aims to enhance synchronization support in Unikraft's internal libraries and port [*mimalloc*](https://github.com/microsoft/mimalloc) (pronounced "me-malloc"), a high-performance general-purpose memory allocator developed by Microsoft, to it.
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+* [Radu Nichita](https://github.com/RaduNichita)
+* [Sergiu Moga](https://github.com/mogasergiu)
+* [Răzvan Deaconescu](https://github.com/razvand)
+
+## Contributions
+
+### Porting *mimalloc*
+
+I started my work based on [Hugo Lefeuvre's](https://github.com/hlef) initial attempt to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)).
+As Unikraft has evolved significantly over the years, I have to adapt the port to work with the latest Unikraft core (v0.17.0), which includes:
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
As Unikraft has evolved significantly over the years, I've had to adapt the port to work with the latest Unikraft core (v0.17.0), which includes:
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    - [PR: Adapt `lib-mimalloc` to using `lib-musl` and the new threading interface](https://github.com/unikraft/lib-mimalloc/pull/12)
+
+- [PR: Adapt Unikraft's heap initialization routine to support *mimalloc*](https://github.com/unikraft/unikraft/pull/1480)
+
+### Benchmarking *mimalloc*
+
+I benchmarked the *mimalloc* allocator using the `cache-scratch` and `cache-thrash` benchmarks *without SMP support enabled*.
+The details of how the benchmarks work can be found in [my second blog post](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft).
+
+In debugging the benchmarks, I have also ran into an confusing bug where the TLS's magic value would be corrupted when the user declares a thread-local array larger than 15 bytes.
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
In debugging the benchmarks, I have also ran into a confusing bug where the TLS's magic value would be corrupted when the user declares a thread-local array larger than 15 bytes.
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+The product of this part of my work includes:
+
+- [Gist: The benchmark application ported to Unikraft](https://gist.github.com/huyang531/28a31fc2d9f348fafb5b9d8e6c9493d5)
+- [Issue: Cannot handle thread-local arrays larger than 15 bytes](https://github.com/unikraft/unikraft/issues/1478)
+
+### Other works
+
+During the course of this project, I have also fixed a few minor bugs and amended Unikraft's document:
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
During the course of this project, I have also fixed a few minor bugs and amended Unikraft's documentation:
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+After running the benchmarks under multiple configurations, we noticed that *mimalloc* did not perform significantly better than the existing *binary buddy* allocator.
+This is because we are not running the virtual machine with SMP enabled, so *mimalloc* would have wasted much time on unnecessary synchronization.
+Therefore, to really benchmark the performance of *mimalloc* (and to justify Unikraft's potential for full SMP support), we need to get it running with real parallelism.
+
+I have outlined some key considerations and future steps in [this blog post](https://github.com/unikraft/docs/pull/455).
+
+### Further testing *mimalloc*
+
+So far, we have only tested *mimalloc* on two multi-threaded benchmarks. To further validate its functionality, portability, and performance. This includes running *mimalloc* on [some more complicated benchmarks](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft#next-steps) and on real-world applications like Redis.
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]: Please split this paragraph on 1 line per sentence. Also, please rephrase `To further validate its functionality, portability, and performance.`, it seems like it is an uncompleted phrase.

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+### Porting the latest *mimalloc* to Unikraft
+
+So far, we have been working on *mimalloc* v1.6.3. We can further improve our work by porting a newer version of the memory allocator to Unikraft.
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]: Same on splitting the paragraph at one row per sentence.

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+### Towards full synchronization support in internal libraries
+
+As the project title suggests, more work needs to be done for Unikraft to provide general synchronization support for user applications, which goes far beyond just providing a synchronous memory allocator. For example, we may need to synchronize the Virtual Memory interface (`ukvmem`), the cooperative scheduler (`ukschedcoop`), and rethink how every component interacts with each other in an SMP environment.
```
Comment (razvanvirtan) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]: Same as above

**Review by razvanvirtan — CHANGES_REQUESTED** [review]
Great job, thanks again for your work!
Once the comments are addressed, we can merge it 🚀 

Reviewed-by : Razvan Virtan <virtanrazvan@gmail.com>

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
+# Unikraft GSoC'24: Synchronization Support In Internal Libraries
+
+<img src="https://unikraft.org/images/unikraft-gsoc24.png"/>
+
+## Summary
+
+Memory allocators have a significant impact on application performance.
+There have been some [research papers](https://dl.acm.org/doi/10.1145/378795.378821) which compared different memory allocators.
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
There have been serveral [research papers](https://dl.acm.org/doi/10.1145/378795.378821) which compared different memory allocators.
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+According to their work, switching to the appropiate memory allocator may improve the application performance by 60%.
+Unikraft currently supports only one non-synchronized memory allocator, the *binary buddy* allocator.
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
Unikraft currently supports only one non-synchronized memory allocator, the [*binary buddy* allocator](https://github.com/unikraft/unikraft/tree/staging/lib/ukallocbbuddy).
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+The details of how the benchmarks work can be found in [my second blog post](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft).
+
+In debugging the benchmarks, I have also ran into a confusing bug where the TLS's magic value would be corrupted when the user declares a thread-local array larger than 15 bytes.
+
+The product of this part of my work includes:
+
+- [Gist: The benchmark application ported to Unikraft](https://gist.github.com/huyang531/28a31fc2d9f348fafb5b9d8e6c9493d5)
+- [Issue: Cannot handle thread-local arrays larger than 15 bytes](https://github.com/unikraft/unikraft/issues/1478)
+
+### Other works
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
### Other Works
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+## Summary
+
+Memory allocators have a significant impact on application performance.
+There have been some [research papers](https://dl.acm.org/doi/10.1145/378795.378821) which compared different memory allocators.
+According to their work, switching to the appropiate memory allocator may improve the application performance by 60%.
+Unikraft currently supports only one non-synchronized memory allocator, the *binary buddy* allocator.
+Therefore, this project aims to enhance synchronization support in Unikraft's internal libraries and port [*mimalloc*](https://github.com/microsoft/mimalloc) (pronounced "me-malloc"), a high-performance general-purpose memory allocator developed by Microsoft, to it.
+
+## GSoC contributor
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
## GSoC Contributor
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+- [PR: Add tips for disabling compiler optimization in debug guide](https://github.com/unikraft/docs/pull/442)
+
+## Blog posts
+
+ - [Blog post #1: Porting *mimalloc*](https://unikraft.org/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft)
+ - [Blog post #2: Testing *mimalloc*](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft)
+ - [Blog post #3: Debugging *mimalloc*](https://github.com/unikraft/docs/pull/448)
+ - [Blog post #4: Benchmarking *mimalloc* in an SMP environment](https://github.com/unikraft/docs/pull/455)
+
+
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion

```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+### Other works
+
+During the course of this project, I have also fixed a few minor bugs and amended Unikraft's documentation:
+
+- [PR: lib/ukvmem: Fix dependency and test scripts](https://github.com/unikraft/unikraft/pull/1447)
+- [PR: Fix typos and add notes about the hardware acceleration with KVM for random number generation](https://github.com/unikraft/docs/pull/452)
+- [PR: Add tips for disabling compiler optimization in debug guide](https://github.com/unikraft/docs/pull/442)
+
+## Blog posts
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
## Blog Posts
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+ - [Blog post #1: Porting *mimalloc*](https://unikraft.org/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft)
+ - [Blog post #2: Testing *mimalloc*](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft)
+ - [Blog post #3: Debugging *mimalloc*](https://github.com/unikraft/docs/pull/448)
+ - [Blog post #4: Benchmarking *mimalloc* in an SMP environment](https://github.com/unikraft/docs/pull/455)
+
+
+## Future work
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
## Future Work
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+
+### Running *mimalloc* in an SMP environment
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
### Running *mimalloc* in an SMP Environment
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+After running the benchmarks under multiple configurations, we noticed that *mimalloc* did not perform significantly better than the existing *binary buddy* allocator.
+This is because we are not running the virtual machine with SMP enabled, so *mimalloc* would have wasted much time on unnecessary synchronization.
+Therefore, to really benchmark the performance of *mimalloc* (and to justify Unikraft's potential for full SMP support), we need to get it running with real parallelism.
+
+I have outlined some key considerations and future steps in [this blog post](https://github.com/unikraft/docs/pull/455).
+
+### Further testing *mimalloc*
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
### Further Testing of *mimalloc*
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+So far, we have only tested *mimalloc* on two multi-threaded benchmarks.
+To further validate its functionality, portability, and performance, we need to test *mimalloc* on [some more complicated benchmarks](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft#next-steps) and on real-world applications like Redis.
+
+### Porting the latest *mimalloc* to Unikraft
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
### Porting the Latest *mimalloc* to Unikraft
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+So far, we have been working on *mimalloc* v1.6.3.
+We can further improve our work by porting a newer version of the memory allocator to Unikraft.
+
+### Towards full synchronization support in internal libraries
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
### Towards Full Synchronization Support in Internal Libraries
```

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+
+As the project title suggests, more work needs to be done for Unikraft to provide general synchronization support for user applications, which goes far beyond just providing a synchronous memory allocator.
+For example, we may need to synchronize the Virtual Memory interface (`ukvmem`), the cooperative scheduler (`ukschedcoop`), and rethink how every component interacts with each other in an SMP environment.
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
For example, we may need to synchronize the [Virtual Memory interface (`ukvmem`)](https://github.com/unikraft/unikraft/tree/staging/lib/ukvmem), the [cooperative scheduler (`ukschedcoop`)](https://github.com/unikraft/unikraft/tree/staging/lib/ukschedcoop), and rethink how every component interacts with each other in an SMP environment.
```

**Review by razvand — CHANGES_REQUESTED** [review]
Thanks, @huyang531 . See my comments.

File: gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+* [Răzvan Deaconescu](https://github.com/razvand)
+
+## Contributions
+
+### Porting *mimalloc*
+
+I started my work based on [Hugo Lefeuvre's](https://github.com/hlef) initial attempt to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)).
+As Unikraft has evolved significantly over the years, I've had to adapt the port to work with the latest Unikraft core (v0.17.0), which includes:
+
+- Switch *mimalloc* from using `newlib` to `MUSL`:
```
Comment (razvand) [inline gsoc-2024/work-product/Yang-Hu-Synchronization-Support-in-Internal-Libraries/Yang-Hu-Synchronization-Support-in-Internal-Libraries.md]:
```suggestion
- Switch *mimalloc* from using `newlib` to `musl`:
```

**Review by razvand — CHANGES_REQUESTED** [review]

**Review by razvand — APPROVED** [review]
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**